cou/steps/analyze.py

```diff
@@ -0,0 +1,488 @@
+        try:
+            pkg_name = CHARM_TYPES[self.charm]["pkg"]
+        except KeyError:
+            logging.debug("package not found for application: %s", self.name)
+            pkg_name = ""
+        return pkg_name
+
+    # NOTE (gabrielcocenza) Ideally, the application should provide the openstack version
+    # and packages versions by a charm action. This might be possible with Sunbeam.
+    def get_current_os_versions(self, unit: str) -> str:
+        """Get the openstack version of a unit."""
+        version = ""
+        pkg_version = get_pkg_version(unit, self.pkg_name, self.model_name)
+        self.units[unit]["pkg_version"] = pkg_version
+        self.pkg_version_units[pkg_version].add(unit)
+
+        # for openstack releases >= wallaby
+        codename = get_openstack_release(unit, model_name=self.model_name)
+        if codename:
+            version = codename
+        # for openstack releases < wallaby
+        elif self.pkg_name and pkg_version:
+            version = get_os_code_info(self.pkg_name, pkg_version)
+        return version
+
+    def get_os_origin(self) -> str:
+        """Get application configuration for openstack-origin or source."""
+        for origin in ("openstack-origin", "source"):
+            if self.config.get(origin):
+                return self.config[origin].get("value", "")
+
+        logging.warning("Failed to get origin for %s, no origin config found", self.name)
+        return ""
+
+    def check_os_versions_units(
+        self, upgrade_units: defaultdict[str, set]
+    ) -> defaultdict[str, set]:
+        """Check openstack versions in an application."""
+        if len(self.os_release_units.keys()) <= 1:
+            return upgrade_units
+
+        logging.warning("Units from %s are not in the same openstack version", self.name)
+        os_sequence = sorted(
+            list(self.os_release_units.keys()),
+            key=CompareOpenStack,
+        )
+        for os_release in os_sequence[:-1]:
+            next_release = determine_next_openstack_release(os_release)[1]
+            upgrade_units[next_release].update(self.os_release_units[os_release])
+            logging.warning(
+                "upgrade units: %s from: %s to %s",
+                self.os_release_units[os_release],
+                os_release,
+                next_release,
+            )
+        return upgrade_units
+
+    def check_os_channels_and_migration(
+        self, change_channel: defaultdict[str, set], charmhub_migration: defaultdict[str, set]
+    ) -> Tuple:
+        """Check openstack channel and if it's necessary a charmhub migration."""
+        if len(self.os_release_units.keys()) > 1:
+            logging.warning(
+                "Skip check of channels. App %s has units with different openstack version",
+                self.name,
+            )
+            return change_channel, charmhub_migration
+
+        actual_release = list(self.os_release_units.keys())[0]
```

Comment:
What if there is no units?

Reply:
If we have an application without units something is wrong with the model, right? I think this is a very rare scenario. Do you think it worth adding a try except IndexError?